src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/Delegation/DesignTimeHelperResponseRewriter.cs

```diff
@@ -59,57 +55,49 @@ internal class DesignTimeHelperResponseRewriter : DelegatedCompletionResponseRew
-        // (e.g. "__ab[||]"), we only trim out common design time helpers from the completion list.
+        // We should remove Razor design-time helpers from C#'s completion list. If the current identifier
+        // being targeted does not start with a double underscore, we trim out all items starting with "__"
+        // from the completion list. If the current identifier does start with a double underscore (e.g. "__ab[||]"),
+        // we only trim out common design time helpers from the completion list.
 
-        var filteredItems = completionList.Items.Except(s_designTimeHelpersCompletionItems, CompletionItemComparer.Instance).ToArray();
+        using var _ = ListPool<CompletionItem>.GetPooledObject(out var filteredItems);
 
-        if (ShouldRemoveAllDesignTimeItems(owner, sourceText))
-        {
-            filteredItems = filteredItems.Where(item => item.Label != null && !item.Label.StartsWith("__", StringComparison.Ordinal)).ToArray();
-        }
+        var items = completionList.Items;
+        filteredItems.SetCapacityIfLarger(items.Length);
 
-        completionList.Items = filteredItems;
-        return completionList;
-    }
+        var shouldRemoveAllDesignTimeItems = StartsWithDoubleUnderscore(owner, sourceText);
 
-    // If the current identifier starts with "__", only trim out common design time helpers from the list.
-    // In all other cases, trim out both common design time helpers and all completion items starting with "__".
-    private static bool ShouldRemoveAllDesignTimeItems(RazorSyntaxNode owner, SourceText sourceText)
-    {
-        if (owner.Span.Length < 2)
+        foreach (var item in items)
         {
-            return true;
+            if (s_designTimeHelpers.Contains(item.Label) || (shouldRemoveAllDesignTimeItems && item.Label.StartsWith("__")))
+            {
+                continue;
+            }
+
+            filteredItems.Add(item);
         }
 
-        if (sourceText[owner.Span.Start] == '_' && sourceText[owner.Span.Start + 1] == '_')
+        // Avoid allocating array if nothing was filtered.
+        if (items.Length != filteredItems.Count)
         {
-            return false;
+            completionList.Items = filteredItems.ToArray();
         }
 
-        return true;
+        return completionList;
     }
 
-    private class CompletionItemComparer : IEqualityComparer<CompletionItem>
+    private static bool StartsWithDoubleUnderscore(RazorSyntaxNode owner, SourceText sourceText)
     {
-        public static CompletionItemComparer Instance = new();
+        // If the current identifier starts with "__", only trim out common design time helpers from the list.
+        // In all other cases, trim out both common design-time helpers and all completion items starting with "__".
 
-        public bool Equals(CompletionItem? x, CompletionItem? y)
+        var span = owner.Span;
+        if (span.Length < 2)
```

Comment:
If you wanted hyper-premature-optimization, could you use `owner.SpanStart`, and check `sourceText.Length < owner.SpanStart +2`?

Reply:
That logic doesn't seem to be quite equivalent, since it allows for a `RazorSyntaxNode` to be of a length less than 2 but still evaluate as starting with a double underscore. Since this is only called once per completion list rather than once per item, I don't think the sacrifice in readability is worth it.